crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -6122,20 +6127,58 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
             );
         }
 
-        let attribute_exists = self
-            .class_context_of_current_method()
-            .and_then(|class| {
-                Type::instance(self.db(), class.default_specialization(self.db()))
-                    .member(self.db(), id)
-                    .place
-                    .ignore_possibly_unbound()
+        let current_method_decorators = self.current_method_decorators();
+
+        let class_context = self.class_context_of_current_method();
+
+        if current_method_decorators.clone().is_some_and(|decorators| {
+            decorators.iter().any(|decorator| {
+                match self.file_expression_type(&decorator.expression) {
+                    Type::ClassLiteral(class) => {
+                        matches!(class.known(self.db()), Some(KnownClass::Classmethod))
+                    }
+                    _ => false,
+                }
             })
-            .is_some();
+        }) {
+            let class_attribute_exists = class_context
+                .and_then(|class| {
+                    SubclassOfType::from(self.db(), class.default_specialization(self.db()))
+                        .member(self.db(), id)
+                        .place
+                        .ignore_possibly_unbound()
+                })
+                .is_some();
+            if class_attribute_exists {
+                diagnostic.info(format_args!(
+                    "An attribute `{id}` is available: consider using `cls.{id}`"
```

Comment:
Not sure if you were intending to leave this for a later PR (if so a test with TODO comment would still be good), but @AlexWaygood 's point in the linked issue was that we should not hardcode either the name `self` or the name `cls` based on what kind of method it is -- rather we should use the actual name of the actual first parameter of the method. It's perfectly valid Python to use a name other than `self` as first parameter of a regular method, or a name other than `cls` as first parameter of a classmethod. (We should also have tests for these cases.)